controller/volume_controller.go

```diff
@@ -1544,3 +1550,93 @@ func (vc *VolumeController) processMigration(v *longhorn.Volume, es map[string]*
 	logrus.Infof("volume %v: migration: migration node %v is ready", v.Name, v.Spec.MigrationNodeID)
 	return nil
 }
+
+func (vc *VolumeController) CheckEngineImageReadiness(image string) error {
+	ei, err := vc.getEngineImage(image)
+	if err != nil {
+		return errors.Wrapf(err, "unable to get engine image %v", image)
+	}
+	if ei.Status.State != types.EngineImageStateReady {
+		return fmt.Errorf("engine image %v (%v) is not ready, it's %v", ei.Name, image, ei.Status.State)
+	}
+	return nil
+}
+
+func (vc *VolumeController) GetEngineClient(volumeName string) (client engineapi.EngineClient, err error) {
+	var e *longhorn.Engine
+
+	defer func() {
+		err = errors.Wrapf(err, "cannot get client for volume %v", volumeName)
+	}()
+	es, err := vc.ds.ListVolumeEngines(volumeName)
+	if err != nil {
+		return nil, err
+	}
+	if len(es) == 0 {
+		return nil, fmt.Errorf("cannot fine engine")
+	}
+	if len(es) != 1 {
+		return nil, fmt.Errorf("more than one engine exists")
+	}
+	for _, e = range es {
+		break
+	}
+	if e.Status.CurrentState != types.InstanceStateRunning {
+		return nil, fmt.Errorf("engine is not running")
+	}
+
+	if err := vc.CheckEngineImageReadiness(e.Spec.EngineImage); err != nil {
+		return nil, errors.Wrapf(err, "cannot get engine client with image %v", e.Spec.EngineImage)
+	}
+
+	engineCollection := &engineapi.EngineCollection{}
+	return engineCollection.NewEngineClient(&engineapi.EngineClientRequest{
+		VolumeName:  e.Spec.VolumeName,
+		EngineImage: e.Spec.EngineImage,
+		IP:          e.Status.IP,
+	})
+}
+
+func (vc *VolumeController) Restore(v *longhorn.Volume, e *longhorn.Engine) error {
+	engine, err := vc.GetEngineClient(v.Name)
+	if err != nil {
+		return fmt.Errorf("failed to get engine client: %v", err)
+	}
+
+	if !v.Spec.RestorationRequired {
+		return nil
+	}
+
+	if v.Spec.FromBackup == "" {
+		return fmt.Errorf("backupURL is not available")
+	}
+
+	if err := engine.BackupRestore(v.Spec.FromBackup); err != nil {
+		logrus.Errorf("Failed to initiate Backup Restore from Volume Controller: %v", err)
+		return err
+	}
+
+	logrus.Infof("Successfully initiated Backup Restore for URL: %v. Checking status on [%v]replicas",
+		v.Spec.FromBackup, len(e.Spec.ReplicaAddressMap))
+
+	for j := 0; j < RetryCounts; j++ {
```

Comment:
See engine PR comment. And you cannot block the volume controller code flow here.

Reply:
This code will only hit when RestorationRequired is true. 
As you know, In ReconcileVolumeState, the following sequence of operation happens on a high level:
1. RestorationRequired set to true
2. Replicas are started
3. Engine is started
4. Restore the replicas
5. Volume is Attached
6. If it's a restoration volume, the Volume is Detached and RestorationRequired is set to False

If the volume is Standby, I understand we can check the restore progress in asynchronous way in engine controller, but if the volume is a Restoration volume, then it has to be a blocking operation right, as it's getting detached as soon as it comes out of the Restore function?

Reply:
For restoration volume, before the restoration completed we should able to see the progress. Since the detach depends on the restoration completion, so the status has already been written to the engine object, we should able to read from it as well.